src/main.c

```diff
@@ -0,0 +1,151 @@
+    osc_init.LSIState = RCC_LSI_ON;
+    osc_init.PLL.PLLState = RCC_PLL_ON;
+    osc_init.PLL.PLLSource = RCC_PLLSOURCE_HSE;
+    osc_init.PLL.PLLM = 6;
+    osc_init.PLL.PLLN = 168;
+    osc_init.PLL.PLLP = RCC_PLLP_DIV2;
+    osc_init.PLL.PLLQ = 7;
+    if (HAL_RCC_OscConfig(&osc_init) != HAL_OK) {
+        Error_Handler();
+    }
+
+    // Initializes the CPU, AHB and APB busses clocks
+    RCC_ClkInitTypeDef clk_init = { 0 };
+    clk_init.ClockType = RCC_CLOCKTYPE_HCLK | RCC_CLOCKTYPE_SYSCLK
+        | RCC_CLOCKTYPE_PCLK1 | RCC_CLOCKTYPE_PCLK2;
+    clk_init.SYSCLKSource = RCC_SYSCLKSOURCE_PLLCLK;
+    clk_init.AHBCLKDivider = RCC_SYSCLK_DIV1;
+    clk_init.APB1CLKDivider = RCC_HCLK_DIV4;
+    clk_init.APB2CLKDivider = RCC_HCLK_DIV2;
+    if (HAL_RCC_ClockConfig(&clk_init, FLASH_LATENCY_5) != HAL_OK) {
+        Error_Handler();
+    }
+
+    RCC_PeriphCLKInitTypeDef periph_clk_init = { 0 };
+    periph_clk_init.PeriphClockSelection = RCC_PERIPHCLK_RTC;
+    periph_clk_init.RTCClockSelection = RCC_RTCCLKSOURCE_LSI;
+    if (HAL_RCCEx_PeriphCLKConfig(&periph_clk_init) != HAL_OK) {
+        Error_Handler();
+    }
+}
+
+static void enable_trap_on_division_by_zero() {
+    SCB->CCR |= SCB_CCR_DIV_0_TRP_Msk;
+}
+
+static void enable_backup_domain() {
+    // this allows us to use the RTC->BKPXX registers
+    __HAL_RCC_PWR_CLK_ENABLE();
+    HAL_PWR_EnableBkUpAccess();
+}
+
+static void enable_segger_sysview() {
+    SEGGER_SYSVIEW_Conf();
+}
+
+static void enable_dfu_entry() {
+#ifdef BUDDY_ENABLE_DFU_ENTRY
+    // check whether user requested to enter the DFU mode
+    // this has to be checked after having
+    //  1) initialized access to the backup domain
+    //  2) having initialized related clocks (SystemClock_Config)
+    if (sys_dfu_requested())
+        sys_dfu_boot_enter();
+#endif
+}
+
+static void eeprom_init_i2c() {
+    // Init the I2C
+    hi2c1.Instance = I2C1;
+    hi2c1.Init.ClockSpeed = 100000;
+    hi2c1.Init.DutyCycle = I2C_DUTYCYCLE_2;
+    hi2c1.Init.OwnAddress1 = 0;
+    hi2c1.Init.AddressingMode = I2C_ADDRESSINGMODE_7BIT;
+    hi2c1.Init.DualAddressMode = I2C_DUALADDRESS_DISABLE;
+    hi2c1.Init.OwnAddress2 = 0;
+    hi2c1.Init.GeneralCallMode = I2C_GENERALCALL_DISABLE;
+    hi2c1.Init.NoStretchMode = I2C_NOSTRETCH_DISABLE;
+    if (HAL_I2C_Init(&hi2c1) != HAL_OK) {
+        Error_Handler();
+    }
+}
+
+/// The entrypoint of the startup task
+///
+/// WARNING
+/// The C++ runtime isn't initialized at the beginning of this function
+/// and initializing it is the main priority here.
+/// So first, we have to get the EEPROM ready, then we call libc_init_array
+/// and that is the time everything is ready for us to switch to C++ context.
+static void startup_task(void const *argument) {
+    // init crc32 module. We need crc in eeprom_init
+    crc32_init();
+
+    // init communication with eeprom
+    eeprom_init_i2c();
+
+    // init eeprom module itself
+    taskENTER_CRITICAL();
+    eeprom_init();
```

Comment:
:thinking: this is walking on thin ice - eeprom_init() is defined inside a C++ source file and may actually call something which is depending on the C++ runtime/globals/constructors. And if not now, someone could add something like that in the future, there is nothing preventing from doing so.